### PR TITLE
Re-introduce Canary network trait

### PR DESCRIPTION
**Pull request by zosorock**
## Motivation

Re-introduces https://github.com/AleoNet/snarkVM/pull/2463

Previously https://github.com/AleoNet/snarkVM/pull/2465 reverted a number of PRs to ensure they can be tested properly. The time has come to test the PRs, so we're doing a partial revert revert.

With https://github.com/AleoNet/snarkVM/pull/2463 back in place they can all be tested on Canary.

## Related PRs
https://github.com/AleoNet/snarkVM/pull/2469 - clone
https://github.com/AleoNet/snarkOS/pull/3282 - sister PR

### CODE REVIEW

File: parameters/src/canary/resources/set_validator_state.metadata [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "e8fbb86008343485bb95c3e3e75d41b493ba654cc161d8b9f9c6901cb71d7a9d",
+  "prover_size": 17390188,
+  "verifier_checksum": "d0942ee642048e21cbf562737b4760524dd378ac17867e7a225736badb23f614",
+  "verifier_size": 673
```
Comment (iamalwaysuncomfortable) [inline parameters/src/canary/resources/set_validator_state.metadata:5]: The parameter scripts should optimally be re-run for both Testnet and Mainnet as well.

File: console/network/src/canary_v0.rs [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    /// The genesis block coinbase target.
+    const GENESIS_COINBASE_TARGET: u64 = (1u64 << 10).saturating_sub(1);
+    /// The genesis block proof target.
+    const GENESIS_PROOF_TARGET: u64 = 1u64 << 8;
+    /// The fixed timestamp of the genesis block.
+    const GENESIS_TIMESTAMP: i64 = 1715776496 /* 2024-05-15 12:34:56 UTC */;
+    /// The network ID.
+    const ID: u16 = 2;
+    /// The function name for the inclusion circuit.
+    const INCLUSION_FUNCTION_NAME: &'static str = MainnetV0::INCLUSION_FUNCTION_NAME;
```
Comment (iamalwaysuncomfortable) [inline console/network/src/canary_v0.rs:144]: This should optimally not reference another concrete trait implementation's name. It should reference its own.

File: console/network/src/canary_v0.rs [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+    }
+
+    /// Returns the Varuna universal prover.
+    fn varuna_universal_prover() -> &'static UniversalProver<Self::PairingCurve> {
+        MainnetV0::varuna_universal_prover()
+    }
+
+    /// Returns the Varuna universal verifier.
+    fn varuna_universal_verifier() -> &'static UniversalVerifier<Self::PairingCurve> {
+        MainnetV0::varuna_universal_verifier()
```
Comment (iamalwaysuncomfortable) [inline console/network/src/canary_v0.rs:217]: Optimally this shouldn't reference MainnetV0

File: parameters/src/canary/mod.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        $map.insert($name.to_string(), std::sync::Arc::new(key));
+    }};
+}
+
+// Inclusion
+impl_remote!(InclusionProver, REMOTE_URL, "resources/", "inclusion", "prover");
+impl_local!(InclusionVerifier, "resources/", "inclusion", "verifier");
+
+/// The function name for the inclusion circuit.
+pub const NETWORK_INCLUSION_FUNCTION_NAME: &str = "inclusion";
```
Comment (iamalwaysuncomfortable) [inline parameters/src/canary/mod.rs:103]: The trait should reference this.

**Review by iamalwaysuncomfortable — COMMENTED** [review]
A few minor things need to be changed about this, highlighted them below.

**Review by vicsn — APPROVED** [review]
This logic was already approved in https://github.com/AleoNet/snarkVM/pull/2463

File: parameters/src/canary/resources/set_validator_state.metadata (hunk 1 above)
Comment (vicsn) [reply parameters/src/canary/resources/set_validator_state.metadata:5]: It works fine as is

File: console/network/src/canary_v0.rs (hunk 2 above)
Comment (vicsn) [reply console/network/src/canary_v0.rs:144]: The PR was already approved, and this is the same methodology used in the testnet trait

File: console/network/src/canary_v0.rs (hunk 3 above)
Comment (vicsn) [reply console/network/src/canary_v0.rs:217]: The PR was already approved, and this is the same methodology used in the testnet trait

File: parameters/src/canary/mod.rs (hunk 4 above)
Comment (vicsn) [reply parameters/src/canary/mod.rs:103]: The PR was already approved, and this is the same methodology used in the testnet trait

**Pull request merged**